core/src/main/java/org/infinispan/util/concurrent/BlockingHandler.java

```diff
@@ -0,0 +1,158 @@
+    * @param <V> stage value type
+    * @return a stage that is complete when the biConsumer is but retains the results from the original stage
+    */
+   <V> CompletionStage<V> whenCompleteBlocking(CompletionStage<V> stage,
+         BiConsumer<? super V, ? super Throwable> biConsumer, Object traceId);
+
+   /**
+    * When the provided stage is complete, continue the completion chain of the returned CompletionStage on the
+    * supplied executor. If tracing is enabled a trace message is printed using the object as an identifier to more
+    * easily track the transition between threads.
+    * <p>
+    * This method is useful when an asynchronous computation completes and you do not want to run further processing
+    * on the thread that returned it. An example may be that some blocking operation is performed on a special blocking
+    * thread pool. However when the blocking operation completes we will want to continue the processing of that result
+    * in a thread pool that is for computational tasks.
+    * <p>
+    * If the supplied stage is already completed when invoking this command, this will return an already completed
+    * stage, which means any additional dependent stages will run in the invoking thread.
+    * @param <V> return value type of the supplied stage
+    * @param delay the stage to delay the continuation until complete
+    * @param traceId the id to print when tracing is enabled
+    * @return a CompletionStage that when depended upon will run any callback in the supplied executor
+    */
+   <V> CompletionStage<V> continueOnNonBlockingThread(CompletionStage<V> delay, Object traceId);
+
+   /**
+    * Provided a publisher that is known to block, this will ensure that the publisher is subscribed on the
+    * blocking thread and any values published will be on a non blocking thread. Note that if a blocking thread
+    * subscribes to the publisher these threads will not be used and thus the entire Publisher is subscribed and
+    * observed on the invoking thread, unless the caller changes this.
+    * @param publisher the publisher that when subscribed to will block
+    * @param <V> the published entry types
+    * @return publisher that will not block the current thread
+    */
+   <V> Publisher<V> blockingPublisher(Publisher<V> publisher);
+
+   /**
+    * Provides a {@link BlockingExecutor} which is limited to the provided concurrency amount.
+    * @param name name of the limited blocking executor
+    * @param concurrency maximum amount of concurrent operations to be performed via the returned executor
+    * @return a blocking executor limited in the amount of concurrent invocations
+    */
+   BlockingExecutor limitedBlockingExecutor(String name, int concurrency);
+
+   /**
+    * Retrieves the non blocking executor for use if needed.
+    * @return executor that runs on non blocking threads
+    */
+   Executor getNonBlockingExecutor();
```

Comment:
-1, it's not such a big deal to inject the non-blocking executor separately. 

Otherwise we shouldn't call it `BlockingHandler` IMO, it would need a name with `Executor` in it instead.